File: Array/zhanchengguo/0630-链表反转/Main.java
package com.test.ktapplication;


import android.util.Log;

public class Main {

    void main() {
        ListNode a = new ListNode(1);
        ListNode b = new ListNode(2);
        ListNode c = new ListNode(3);
        ListNode d = new ListNode(4);
        ListNode e = new ListNode(5);
        a.next = b;
        b.next = c;
        c.next = d;
        d.next = e;
        e.next = null;
        ListNode result = reverseList1(a);
        while (result !=null){
            Log.e("value is ",result.val+"");
            result = result.next;
        }
    }


    /**
     * 递归
     * @param head
     * @return
     */
    public ListNode reverseList1(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        ListNode lastNode = reverseList1(head.next);
        head.next.next = head;
        head.next = null;
        return lastNode;
    }

    /**
     * 双指针
     * @param head
     * @return
     */
    public ListNode reverseList2(ListNode head) {
        ListNode pre = null;
        ListNode current = head;
        while (current != null) {
            ListNode temp = current.next;
            current.next = pre;
            pre = current;
            current = temp;
        }
        return pre;
    }
}

class ListNode {
    int val;
    ListNode next;

    ListNode(int x) {
        val = x;
    }
}

File: Array/zhanchengguo/0630-链表反转/main.kt
package com.test.ktapplication

fun main() {
    val a = ListNode(1)
    val b = ListNode(2)
    val c = ListNode(3)
    val d = ListNode(4)
    val e = ListNode(5)
    a.next = b
    b.next = c
    c.next = d
    d.next = e
    e.next = null
    var result = reverseList1(a)
    while (result != null) {
        print("value is ${result.`val`}\n")
        result = result.next
    }

}

/**
 * 递归
 */
fun reverseList1(head: ListNode?): ListNode? {
    if (head == null || head.next == null) {
        return head
    }
    val lastNode = reverseList1(head.next)
    head?.next?.next = head
    head.next = null
    return lastNode
}

/**
 * 双指针
 */
fun reverseList2(head: ListNode?): ListNode? {
    var pre: ListNode? = null
    var current: ListNode? = head
    while (current != null) {
        var temp = current.next
        current.next = pre
        pre = current
        current = temp
    }
    return pre
}


data class ListNode(var `val`: Int) {
    var next: ListNode? = null
}